### PR TITLE
[Feature] [Integration] [Snyk] - New Snyk Integration

### LINKED ISSUE #88: [Feature] [Integration] [Snyk] - New Snyk Integration
### Motivation

[Snyk](https://snyk.io) is one of the leading tools to scan package vulnerabilities.

A developer portal can leverage having a unified view of the services and their package's vulnerabilities.

Currently, you can achieve that by using Port's custom webhook feature as described [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/webhook/examples/snyk/), but it cannot bring history data about those vulnerabilities.

Therefore, it would be great to implement an Ocean integration that will help to bring data and to map into blueprints in the catalog.

### Proposed Changes

# Files
We will add a new Ocean integration under the `integrations` folder named `snyk`

# Configuration
It's configuration would be as the following:
```yaml
...
integration:
  # The identifier of this integration instance.
  identifier: my-snyk-integration
  # The type of the integration.
  type: "snyk"
  config:
    apiToken: REQUIRED
    webhookSecret: OPTIONAL
    apiUrl: OPTIONAL with default to the default Snyk API url
    organizationId: REQUIRED
    appHost: OPTIONAL
```

The `webhookSecret` +  `appHost` are optional because we want to make event real-time optional so users won't have to configure a webhook URL public endpoint.

# API
Use the latest Snyk API, 2023-08-21 or 2023-08-21~beta.

# Blueprints

## Preview
image

## Blueprints JSON
```json
[
  {
    "identifier": "snykProject",
    "description": "This blueprint represents a snyk project in our software catalog",
    "title": "Snyk Project",
    "icon": "Snyk",
    "schema": {
      "properties": {
        "url": {
          "type": "string",
          "title": "URL",
          "format": "url",
          "icon": "Snyk"
        },
        "owner": {
          "type": "string",
          "title": "Owner",
          "format": "user",
          "icon": "TwoUsers"
        },
        "businessCriticality": {
          "title": "Business Criticality",
          "type": "array",
          "items": {
            "type": "string",
            "enum": [
              "Critical",
              "High",
              "Medium",
              "Low"
            ]
          },
          "icon": "DefaultProperty"
        },
        "environment": {
          "items": {
            "type": "string"
          },
          "icon": "Environment",
          "title": "Environment",
          "type": "array"
        },
        "lifeCycle": {
          "title": "Life Cycle",
          "type": "array",
          "items": {
            "type": "string",
            "enum": [
              "Development",
              "Sandbox",
              "Production"
            ]
          },
          "icon": "DefaultProperty"
        },
        "highOpenVulnerabilities": {
          "icon": "Vulnerability",
          "type": "number",
          "title": "Open High Vulnerabilities"
        },
        "mediumOpenVulnerabilities": {
          "icon": "Vulnerability",
          "type": "number",
          "title": "Open Medium Vulnerabilities"
        },
        "lowOpenVulnerabilities": {
          "icon": "Vulnerability",
          "type": "number",
          "title": "Open Low Vulnerabilities"
        },
        "importedBy": {
          "icon": "TwoUsers",
          "type": "string",
          "title": "Imported By",
          "format": "user"
        },
        "tags": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "title": "Tags",
          "icon": "DefaultProperty"
        },
        "targetOrigin": {
          "title": "Target Origin",
          "type": "string",
          "enum": [
            "artifactory-cr",
            "aws-config",
            "aws-lambda",
            "azure-functions",
            "azure-repos",
            "bitbucket-cloud",
            "bitbucket-server",
            "cli",
            "cloud-foundry",
            "digitalocean-cr",
            "docker-hub",
            "ecr",
            "gcr",
            "github",
            "github-cr",
            "github-enterprise",
            "gitlab",
            "gitlab-cr",
            "google-artifact-cr",
            "harbor-cr",
            "heroku",
            "ibm-cloud",
            "kubernetes",
            "nexus-cr",
            "pivotal",
            "quay-cr",
            "terraform-cloud"
          ],
          "icon": "DefaultProperty"
        },
        "targetUrl": {
          "title": "Target URL",
          "type": "string",
          "format": "url",
          "icon": "Snyk"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "relations": {}
  },
  {
    "identifier": "snykVulnerability",
    "description": "This blueprint represents a Snyk vulnerability in our software catalog",
    "title": "Snyk Vulnerability",
    "icon": "Snyk",
    "schema": {
      "properties": {
        "score": {
          "icon": "Star",
          "type": "number",
          "title": "Score"
        },
        "packageName": {
          "type": "string",
          "title": "Package Name",
          "icon": "DefaultProperty"
        },
        "packageVersions": {
          "icon": "Package",
          "title": "Package Versions",
          "type": "array"
        },
        "type": {
          "type": "string",
          "title": "Type",
          "enum": [
            "vuln",
            "license",
            "configuration"
          ],
          "icon": "DefaultProperty"
        },
        "severity": {
          "icon": "Alert",
          "title": "Issue Severity",
          "type": "string",
          "enum": [
            "low",
            "medium",
            "high",
            "critical"
          ],
          "enumColors": {
            "low": "green",
            "medium": "yellow",
            "high": "red",
            "critical": "red"
          }
        },
        "url": {
          "icon": "Link",
          "type": "string",
          "title": "Issue URL",
          "format": "url"
        },
        "language": {
          "type": "string",
          "title": "Language",
          "icon": "DefaultProperty"
        },
        "publicationTime": {
          "type": "string",
          "format": "date-time",
          "title": "Publication Time",
          "icon": "DefaultProperty"
        },
        "isPatched": {
          "type": "boolean",
          "title": "Is Patched",
          "icon": "DefaultProperty"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "relations": {
      "snykProject": {
        "title": "Project",
        "target": "snykProject",
        "required": false,
        "many": false
      }
    }
  }
]
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by danielsinai**
Snyk integration
Fixes #88

### CODE REVIEW

File: integrations/snyk/README.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+	--set integration.secrets.token="<your-token>"  \
+  --set integration.config.organizationId="<your-organization-id>"  \
+  --set ingress.enabled=true  \
+  --set ingress.annotations."nginx\.ingress\.kubernetes\.io/rewrite-target"= / 
+```
+
+## Supported Kinds
+
+### Target
+This kind represents a Snyk target. The schema should be similar to the one on the [Snyk REST API documentation](https://apidocs.snyk.io/?version=2023-08-29%7Ebeta#tag--Targets).
```
Comment (Tankilevitch) [inline integrations/snyk/README.md:39]: we need to state what snyk api version the integration supports

File: integrations/snyk/snyk/client.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        webhook_secret: str | None,
+    ):
+        self.token = token
+        self.api_url = f"{api_url}/v1"
+        self.app_host = app_host
+        self.org_id = org_id
+        self.rest_api_url = f"{api_url}/rest"
+        self.webhook_secret = webhook_secret
+        self.http_client = httpx.AsyncClient(headers=self.api_auth_header, timeout=30)
+        self.snyk_api_version = "2023-08-21"
```
Comment (Tankilevitch) [inline integrations/snyk/snyk/client.py:24]: do we rely on the api version or this can be configured by the user? (in the azure exporter we allow the user to specify the api-version per resource kind )

File: integrations/snyk/snyk/client.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                [
+                    project
+                    for project in all_projects
+                    if project.get("__target", {}).get("data", {}).get("id")
+                    == target_id
+                ]
+                if target_id
+                else all_projects
+            )
+            yield projects_to_yield
```
Comment (Tankilevitch) [inline integrations/snyk/snyk/client.py:151]: you repeat the same code from above, lets reuse code

File: integrations/snyk/snyk/client.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+    async def get_single_target_by_project_id(self, project_id: str) -> dict[str, Any]:
+        project = await self.get_single_project({"id": project_id})
+        target_id = project.get("__target", {}).get("data", {}).get("id")
+
+        url = f"{self.rest_api_url}/orgs/{self.org_id}/targets/{target_id}"
+        query_params = {"version": f"{self.snyk_api_version}~beta"}
```
Comment (Tankilevitch) [inline integrations/snyk/snyk/client.py:158]: why beta? I think we want to support only GA versions

Comment (Tankilevitch) [reply integrations/snyk/snyk/client.py:158]: I think this needs to be configurable either way

File: integrations/snyk/snyk/client.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        return user_details
+
+    async def get_cached_target_details(self, target_id: str) -> dict[str, Any]:
+        cached_details = event.attributes.get(target_id)
+        if cached_details:
+            return cached_details
+
+        target_details = await self._send_api_request(
+            url=f"{self.rest_api_url}/orgs/{self.org_id}/targets/{target_id}",
+            query_params={"version": f"{self.snyk_api_version}~beta"},
```
Comment (Tankilevitch) [inline integrations/snyk/snyk/client.py:224]: same

File: integrations/snyk/snyk/client.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            except httpx.HTTPStatusError as e:
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+                )
+                raise
+
+    async def get_issues(self, project: dict[str, Any]) -> list[dict[str, Any]]:
+        def build_cache_key(project_id: str) -> str:
+            return f"issues-{project_id}"
+
```
Comment (Tankilevitch) [inline integrations/snyk/snyk/client.py:83]: I would add a comment that we cache the issues for the resync mechanism as there are multiple resync kinds that needs the issues per project and the resync has the same event context for all resync methods

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Nice Job @PeyGis @danielsinai 🚀 
Add minor comments

File: integrations/snyk/config.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  type: POLLING
+integration:
+  # The identifier of this integration instance.
+  identifier: my-snyk-integration
+  # The type of the integration.
+  type: "snyk"
+  config:
+    token: "{{ from env SNYK_API_TOKEN }}"
+    webhookSecret: "{{ from env SNYK_WEBHOOK_SECRET }}"
+    appHost: "{{ from env LOCAL_TUNNEL }}"
```
Comment (yairsimantov20) [inline integrations/snyk/config.yaml:18]: remove any unrequired field

File: integrations/snyk/main.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        logger.info("Gathered all project issues of projects in batch")
+        all_issues.extend(
+            [
+                issue
+                for project_issues in project_issues_list
+                for issue in project_issues
+            ]
+        )
+
+    return list({issue["id"]: issue for issue in all_issues}.values())
```
Comment (yairsimantov20) [inline integrations/snyk/main.py:88]: why do u care for duplicates?

File: integrations/snyk/main.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        tasks = [process_project(semaphore, project) for project in projects]
+        project_issues_list = await asyncio.gather(*tasks)
+        logger.info("Gathered all project issues of projects in batch")
+        all_issues.extend(
+            [
+                issue
+                for project_issues in project_issues_list
+                for issue in project_issues
+            ]
+        )
```
Comment (yairsimantov20) [inline integrations/snyk/main.py:86]: you can use sum

File: integrations/snyk/main.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+        snyk_client = SnykClient(
+            ocean.integration_config["token"],
+            ocean.integration_config["api_url"],
+            ocean.integration_config.get("app_host"),
+            ocean.integration_config["organization_id"],
+            ocean.integration_config.get("webhook_secret"),
+        )
+
+        project = data.get("project", {})
```
Comment (yairsimantov20) [inline integrations/snyk/main.py:107]: why r u using get?
u already know itll be there

File: integrations/snyk/main.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        project_details = await snyk_client.get_single_project(project)
+        await ocean.register_raw(
+            ObjectKind.ISSUE, await snyk_client.get_issues(project)
+        )
+        await ocean.register_raw(ObjectKind.PROJECT, [project_details])
+        await ocean.register_raw(
+            ObjectKind.TARGET,
+            [await snyk_client.get_single_target_by_project_id(project["id"])],
+        )
```
Comment (yairsimantov20) [inline integrations/snyk/main.py:116]: run them at the same time

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]

File: port_ocean/core/defaults/initialize.py [hunk 12]
```diff
@@ -130,10 +130,8 @@ async def _initialize_defaults(
         )
         await asyncio.gather(
             *(
-                port_client.delete_blueprint(
-                    blueprint["identifier"], should_raise=False
-                )
-                for blueprint in defaults.blueprints
+                port_client.delete_blueprint(identifier, should_raise=False)
+                for identifier in e.blueprints_to_rollback
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:134]: lets move those to a different pr so ull be able to ship this integration with them in the first version without releasing a new one

File: integrations/snyk/.port/spec.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
+version: v0.1.0
+type: snyk
+description: Snyk integration for Port Ocean
+icon: Snyk
+features:
+  - type: exporter
+    section: Code Quality & Security
+    resources:
```
Comment (yairsimantov20) [inline integrations/snyk/.port/spec.yaml:8]: i think u r missing something

File: integrations/snyk/.port/spec.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
+version: v0.1.0
```
Comment (yairsimantov20) [inline integrations/snyk/.port/spec.yaml:1]: this is no longer ued

File: integrations/snyk/main.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        ocean.integration_config["api_url"],
+        ocean.integration_config.get("app_host"),
+        ocean.integration_config["organization_id"],
+        ocean.integration_config.get("webhook_secret"),
+    )
+
+    async for projects in snyk_client.get_paginated_projects():
+        logger.info(f"Received batch with {len(projects)} projects")
+        for project in projects:
+            yield [{**project, "__issues": await snyk_client.get_issues(project)}]
```
[the remark this answers is not in the document]
Comment (danielsinai) [reply integrations/snyk/main.py:48]: Summary
Make the get_issues to run in parallel

File: integrations/snyk/snyk/client.py (hunk 2 above)
Comment (yairsimantov20) [reply integrations/snyk/snyk/client.py:24]: if its something that isnt realy changing i think itll be good enough for now IMO

File: integrations/snyk/snyk/client.py [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def _get_paginated_resources(
+        self,
+        base_url: str,
+        url_path: str,
+        method: str = "GET",
+        data_key: str = "data",
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:55]: remove thalt data key parameter and just use the key in the place that is using it

File: integrations/snyk/snyk/client.py [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+                url_path = data.get("links", {}).get("next")
+            except httpx.HTTPStatusError as e:
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+                )
+                raise
+
+    async def get_issues(self, project: dict[str, Any]) -> list[dict[str, Any]]:
+        def build_cache_key(project_id: str) -> str:
+            return f"issues-{project_id}"
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:82]: no need for that function
seems like the value will always be the same

File: integrations/snyk/snyk/client.py [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+
+        if event.attributes.get(build_cache_key(project["id"])):
+            return event.attributes[build_cache_key(project["id"])]
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:85]:
```suggestion
        if cache:= event.attributes.get(build_cache_key(project["id"])):
            return cache
```

File: integrations/snyk/snyk/client.py [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+
+                # Check if there is a "next" URL in the links object
+                url_path = data.get("links", {}).get("next")
+            except httpx.HTTPStatusError as e:
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+                )
+                raise
+
+    async def get_issues(self, project: dict[str, Any]) -> list[dict[str, Any]]:
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:80]: u r only using the project id then just pass it to this function

File: integrations/snyk/snyk/client.py [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+        def build_cache_key(project_id: str) -> str:
+            return f"issues-{project_id}"
+
+        if event.attributes.get(build_cache_key(project["id"])):
+            return event.attributes[build_cache_key(project["id"])]
+
+        url = f"{self.api_url}/org/{self.org_id}/project/{project.get('id')}/aggregated-issues"
+        try:
+            issues = (
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:89]: isnt it paginated?

File: integrations/snyk/snyk/client.py [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+                    query_params={"version": self.snyk_api_version},
+                )
+            )["issues"]
+            event.attributes[build_cache_key(project["id"])] = issues
+            return issues
+        except Exception as e:
+            logger.error(
+                f"Error fetching issues for project: {project.get('id')} error: {e}"
+            )
+            raise
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:102]: arent u already logging inside?

File: integrations/snyk/snyk/client.py [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+
+    async def get_paginated_projects(
+        self,
+        target_id: Optional[str] = None,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        if "projects" in event.attributes:
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:108]: store the cache keys in consts

File: integrations/snyk/snyk/client.py [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def _get_paginated_resources(
+        self,
+        base_url: str,
+        url_path: str,
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:53]: merge those 2 to one var

File: integrations/snyk/snyk/client.py [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        async for projects in self._get_paginated_resources(
+            base_url=self.rest_api_url, url_path=url, query_params=query_params
+        ):
+            all_projects = []
+
+            for project in projects:
+                enriched_project = await self.enrich_project(project)
+                all_projects.append(enriched_project)
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:137]: can u run this together and use asyncio gather?

File: integrations/snyk/snyk/client.py [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+
+            event.attributes.setdefault("projects", []).extend(all_projects)
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:139]: why dont u set the cache at the end?

File: integrations/snyk/snyk/client.py [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+            projects_to_yield = (
+                [
+                    project
+                    for project in all_projects
+                    if project.get("__target", {}).get("data", {}).get("id")
+                    == target_id
+                ]
+                if target_id
+                else all_projects
+            )
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:150]: maybe this should be a private func

File: integrations/snyk/snyk/client.py [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+            yield projects_to_yield
+
+    async def get_single_target_by_project_id(self, project_id: str) -> dict[str, Any]:
+        project = await self.get_single_project({"id": project_id})
+        target_id = project.get("__target", {}).get("data", {}).get("id")
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:155]: is it possible to not have the target id?

File: integrations/snyk/snyk/client.py [hunk 28]
```diff
@@ -0,0 +1,268 @@
...
+
+        url = f"{self.rest_api_url}/orgs/{self.org_id}/targets/{target_id}"
+        query_params = {"version": f"{self.snyk_api_version}~beta"}
+
+        response = await self._send_api_request(
+            url=url, method="GET", query_params=query_params
+        )
+
+        target = response.get("data", {})
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:164]: is it possible to not have the data key?

File: integrations/snyk/snyk/client.py [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+        async for projects_data_of_target in self.get_paginated_projects(
+            target.get("id")
+        ):
+            target.setdefault("__projects", []).extend(projects_data_of_target)
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:168]: why set default? does the api already return this key?

File: integrations/snyk/snyk/client.py [hunk 30]
```diff
@@ -0,0 +1,268 @@
...
+        return target
+
+    async def get_paginated_targets(
+        self,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"/orgs/{self.org_id}/targets"
+        query_params = {"version": f"{self.snyk_api_version}~beta"}
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:175]: u r doing this over and over

File: integrations/snyk/snyk/client.py [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+            targets_with_project_data = []
+            for target_data in targets:
+                async for projects_data_of_target in self.get_paginated_projects(
+                    target_data.get("id")
+                ):
+                    target_data.setdefault("__projects", []).extend(
+                        projects_data_of_target
+                    )
+                targets_with_project_data.append(target_data)
+            yield targets_with_project_data
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:188]: think about this function again 
i think it can be more parallel

File: integrations/snyk/snyk/client.py [hunk 32]
```diff
@@ -0,0 +1,268 @@
...
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"/orgs/{self.org_id}/targets"
+        query_params = {"version": f"{self.snyk_api_version}~beta"}
+        async for targets in self._get_paginated_resources(
+            base_url=self.rest_api_url, url_path=url, query_params=query_params
+        ):
+            targets_with_project_data = []
+            for target_data in targets:
+                async for projects_data_of_target in self.get_paginated_projects(
+                    target_data.get("id")
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:182]: is it an option not have the id key?

File: integrations/snyk/snyk/client.py [hunk 33]
```diff
@@ -0,0 +1,268 @@
...
+                ):
+                    target_data.setdefault("__projects", []).extend(
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:184]: is it possible that the target already have __projects?

File: integrations/snyk/snyk/client.py [hunk 34]
```diff
@@ -0,0 +1,268 @@
...
+                        projects_data_of_target
+                    )
+                targets_with_project_data.append(target_data)
+            yield targets_with_project_data
+
+    async def get_single_project(self, project: dict[str, Any]) -> dict[str, Any]:
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:190]: u r only using the project id

Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:190]: u might have it in the cache
or you can store it in the cache if its not there

File: integrations/snyk/snyk/client.py [hunk 35]
```diff
@@ -0,0 +1,268 @@
...
+            "expand": "target",
+        }
+        response = await self._send_api_request(
+            url=url, method="GET", query_params=query_params
+        )
+        return await self.enrich_project(response.get("data", {}))
+
+    async def get_cached_user_details(self, user_id: str) -> dict[str, Any]:
+        if (
+            not user_id
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:204]: will the user id be none or an empty string

File: integrations/snyk/snyk/client.py [hunk 36]
```diff
@@ -0,0 +1,268 @@
...
+        ):  ## Some projects may not have been assigned to any owner yet. In this instance, we can return an empty dict
+            return {}
+        cached_details = event.attributes.get(user_id)
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:207]: cache the data under a proper keys that explain what is that data

File: integrations/snyk/snyk/client.py [hunk 37]
```diff
@@ -0,0 +1,268 @@
...
+        if cached_details:
+            return cached_details
+
+        user_details = await self._send_api_request(
+            url=f"{self.api_url}/user/{user_id}"
+        )
+        event.attributes[user_id] = user_details
+        return user_details
+
+    async def get_cached_target_details(self, target_id: str) -> dict[str, Any]:
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:217]: dont use get `cached` in your function name because the data might not be cached

File: integrations/snyk/snyk/client.py [hunk 38]
```diff
@@ -0,0 +1,268 @@
...
+            return cached_details
+
+        target_details = await self._send_api_request(
+            url=f"{self.rest_api_url}/orgs/{self.org_id}/targets/{target_id}",
+            query_params={"version": f"{self.snyk_api_version}~beta"},
+        )
+        event.attributes[target_id] = target_details
+        return target_details
+
+    async def enrich_project(self, project: dict[str, Any]) -> dict[str, Any]:
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:229]: maybe u should load the project into a pydantic model to get the data in a more programmatic way

File: integrations/snyk/snyk/client.py [hunk 39]
```diff
@@ -0,0 +1,268 @@
...
+            .get("importer", {})
+            .get("data", {})
+            .get("id")
+        )
+        target_id = (
+            project.get("relationships", {}).get("target", {}).get("data", {}).get("id")
+        )
+        owner_details = await self.get_cached_user_details(owner_id)
+        importer_details = await self.get_cached_user_details(importer_id)
+        target_details = await self.get_cached_target_details(target_id)
```
Comment (yairsimantov20) [inline integrations/snyk/snyk/client.py:244]: try to parallel as much as possible

File: integrations/snyk/snyk/client.py (hunk 20 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:89]: Nope :((

File: integrations/snyk/snyk/client.py (hunk 18 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:85]: Noice!

File: integrations/snyk/snyk/client.py (hunk 17 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:82]: It's an helper, instead of writing the formatting twice, don't you think it's a bit cleaner?

File: integrations/snyk/snyk/client.py (hunk 33 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:184]: Yeap, because I am doing it in generator inside a generator

File: integrations/snyk/snyk/client.py (hunk 31 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:188]: It might, but it's pretty fast, I think we can use it like this

File: integrations/snyk/snyk/client.py (hunk 34 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:190]: Nice catch!

File: integrations/snyk/snyk/client.py (hunk 35 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:204]: There are project's without a user, so we handle it here

File: integrations/snyk/snyk/client.py (hunk 38 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py:229]: I choose not to, because describing it can take me a while, and also we will need to maintain it.
LMK how urgent you feel it is

File: integrations/snyk/snyk/client.py (hunk 17 above)
Comment (yairsimantov20) [reply integrations/snyk/snyk/client.py]: no need to build multiple times
just set it to a var

File: integrations/snyk/snyk/client.py (hunk 24 above)
Comment (yairsimantov20) [reply integrations/snyk/snyk/client.py:149]: what redundent requests?

**Review by yairsimantov20 — APPROVED** [review]
🌊 Dont forget to bump to the upcoming framework version
Beside that looks good to me

File: integrations/snyk/snyk/client.py (hunk 17 above)
Comment (danielsinai) [reply integrations/snyk/snyk/client.py]: Eh, you are right

**Review by Tankilevitch — APPROVED** [review]
LGTM
